rustls/src/dns_name.rs

```diff
@@ -0,0 +1,176 @@
+/// DNS name validation according to RFC1035, but with underscores allowed.
+use std::error::Error as StdError;
+use std::fmt;
+
+/// "Labels must be 63 characters or less."
+const MAX_LABEL_LENGTH: usize = 63;
+
+fn validate(input: &[u8]) -> Result<(), InvalidDnsNameError> {
+    let mut label_length = 0;
+    let mut previous_label_ch = b'-';
```

Comment:
For both of these, I would consider making them `Option`, having them be reset to `None` once a `.` rolls around.

Should we also error on invalid label lengths (64 characters or more)?

Reply:
Also I like to write these kinds of little parsers more like this:

```rust
fn validate(input: &[u8]) -> Result<(), InvalidDnsNameError> {
    use State::*;
    let mut state = Start;

    for ch in input {
        state = match (state, ch) {
            (Start | Next | Hyphen { .. }, b'.') => return Err(InvalidDnsNameError),
            (Subsequent { .. }, b'.') => Next,
            (Subsequent { len }, _) if len >= MAX_LABEL_LENGTH => return Err(InvalidDnsNameError),
            (Start | Next, b'a'..=b'z' | b'A'..=b'Z') => Subsequent { len: 1 },
            (Subsequent { len }, b'-') => Hyphen { len: len + 1 },
            (
                Subsequent { len } | Hyphen { len },
                b'a'..=b'z' | b'A'..=b'Z' | b'_' | b'0'..=b'9',
            ) => Subsequent { len: len + 1 },
            _ => return Err(InvalidDnsNameError),
        };
    }

    if matches!(state, Start | Hyphen { .. }) {
        return Err(InvalidDnsNameError);
    }

    Ok(())
}

#[derive(Debug)]
enum State {
    Start,
    Next,
    Subsequent { len: usize },
    Hyphen { len: usize },
}
```

What do you think? This makes it a bit more explicit what the state is (less potential for `label_length` and `previous_label_ch` to be in inconsistent or invalid states) and is a bit more compact.

Reply:
Oh, that is a big improvement! Will integrate that.

Reply:
That's a nice refactor :+1: